libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/ViewportDataSourceProcessor.kt

```diff
@@ -0,0 +1,328 @@
+package com.mapbox.navigation.ui.maps.camera.data
+
+import android.util.Log
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.core.constants.Constants
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.maps.EdgeInsets
+import com.mapbox.maps.ScreenCoordinate
+import com.mapbox.maps.Size
+import com.mapbox.navigation.base.trip.model.RouteLegProgress
+import com.mapbox.navigation.base.trip.model.RouteStepProgress
+import com.mapbox.navigation.ui.maps.camera.utils.metersToKilometers
+import com.mapbox.navigation.ui.maps.camera.utils.shortestRotation
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfException
+import com.mapbox.turf.TurfMeasurement
+import com.mapbox.turf.TurfMisc
+import kotlin.math.abs
+
+private const val ZERO_PITCH = 0.0
+
+internal object ViewportDataSourceProcessor {
+
+    /**
+     * Returns complete route points in nested arrays of points for all steps in all legs arranged as \[legs]\[steps]\[points].
+     */
+    fun processRoutePoints(route: DirectionsRoute): List<List<List<Point>>> {
+        return route.legs()?.map { routeLeg ->
+            routeLeg.steps()?.map { legStep ->
+                legStep.geometry()?.let { geometry ->
+                    PolylineUtils.decode(geometry, route.precision()).toList()
+                } ?: emptyList()
+            } ?: emptyList()
+        } ?: emptyList()
+    }
+
+    /**
+     * Returns a list of points that trails each maneuver based on the method arguments.
+     *
+     * Those points can later be used as additional input for framing.
+     */
+    fun processRouteForPostManeuverFramingGeometry(
```

Comment:
+ tests. There's a lot of logic here that we'll break if we change it

Reply:
Added!